src/SubscriberBase.h

```diff
@@ -18,34 +30,116 @@ class SubscriberBaseT : public Subscriber<T>,
+  // for the users of the SubscriberBase
+  class SubscriptionShim : public Subscription {
+   public:
+    explicit SubscriptionShim(
+        std::shared_ptr<SubscriberBaseT<T>> parentSubscriber)
+        : parentSubscriber_(std::move(parentSubscriber)) {}
+
+    void request(size_t n) override final {
+      if (auto parent = parentSubscriber_.lock()) {
+        parent->runInExecutor([parent, n]() {
+          if (!parent->cancelled_) {
+            parent->originalSubscription_->request(n);
+          }
+        });
+      }
+    }
+
+    void cancel() override final {
+      if (auto parent = parentSubscriber_.lock()) {
+        if (!parent->cancelled_.exchange(true)) {
+          parent->runInExecutor([parent]() {
+            if (!parent->cancelled_) {
+              parent->originalSubscription_->cancel();
+              parent->originalSubscription_ = nullptr;
+            }
+          });
+        }
+      }
+    }
+
+   private:
+    // we don't want to create cycle with parent subscriber. If we do, we would
+    // have to make sure the class deriving from SubscriberBase would have to
+    // nullify subscription pointer after calling cancel. That is too strong of
+    // a requirement for the users.
+    std::weak_ptr<SubscriberBaseT<T>> parentSubscriber_;
+  };
+
+  friend class SubscriptionShim;
+
  public:
-  using ExecutorBase::ExecutorBase;
+  // in c++11 we have to declare this explicitly, instead of
+  // using ExecutorBase::ExecutorBase because of atomic cancelled :(
+  // maybe its gcc issue
+  explicit SubscriberBaseT(
+      folly::Executor& executor = defaultExecutor(),
+      bool startExecutor = true)
+      : ExecutorBase(executor, startExecutor), cancelled_(false) {}
 
   void onSubscribe(std::shared_ptr<Subscription> subscription) override final {
     auto thisPtr = this->shared_from_this();
-    runInExecutor(
-        [thisPtr, subscription]() { thisPtr->onSubscribeImpl(subscription); });
+    runInExecutor([thisPtr, subscription]() {
+      CHECK(!thisPtr->cancelled_);
+      CHECK(!thisPtr->originalSubscription_);
+      thisPtr->originalSubscription_ = std::move(subscription);
+      thisPtr->onSubscribeImpl(
+          std::make_shared<SubscriptionShim>(thisPtr->shared_from_this()));
+    });
   }
 
   void onNext(T payload) override final {
     auto movedPayload = folly::makeMoveWrapper(std::move(payload));
     auto thisPtr = this->shared_from_this();
     runInExecutor([thisPtr, movedPayload]() mutable {
-      thisPtr->onNextImpl(movedPayload.move());
+      if (!thisPtr->cancelled_) {
+        thisPtr->onNextImpl(movedPayload.move());
+      }
     });
   }
 
   void onComplete() override final {
     auto thisPtr = this->shared_from_this();
-    runInExecutor([thisPtr]() { thisPtr->onCompleteImpl(); });
+    runInExecutor([thisPtr]() {
+      if (!thisPtr->cancelled_.exchange(true)) {
+        thisPtr->onCompleteImpl();
+
+        DCHECK(thisPtr->originalSubscription_);
```

Comment:
This seems like it should be either a CHECK (to avoid segfault below) or add an IF below